### ISSUE TITLE
Using Object Literals as Const for types in arcgis-rest-js over enums

### ISSUE THREAD

**Issue by subgan82** [Esri/arcgis-rest-js#668]
We had updated to use `enums` in https://github.com/Esri/arcgis-rest-js/issues/654 along with string literals union types. 

This resulted in having both `enum` and `string literals union types` in the latest release.

We see benefits of using `Object Literal as const` over enum as it eliminates the need to have both the enum and string literal union types as seen using enums. This new alternative is backwards compatible and keeps the doc more cleaner. Also using this approach we can iterate over values in runtime which is not possible using `const enums`.

----

> It is boiling down to how badly we need the capability to produce an array of all possible string values. If it is needed, as const is looking more plausible.

Ok , another benefit when using the Object Literals is that we don't have both string union types and enums listed in the types as we have with merged PR.

With Object Literal `as const` approach
`IFont.style` becomes `FontStyle`

```javascript

const FontStyle = {
  Italic: "italic",
  Normal: "normal",
  Oblique: "oblique"
} as const;

export type FontStyle = (typeof FontStyle)[keyof typeof FontStyle];

export interface IFont {
  family?: string; // "<fontFamily>";
  size?: number; // <fontSize>;
  style?: FontStyle;
  weight?: FontWeight;
  decoration?: FontDecoration;
}
```

Merged PR using `const enum`

```javascript
export interface IFont {
  family?: string; // "<fontFamily>";
  size?: number; // <fontSize>;
  style?: "italic" | "normal" | "oblique" | FontStyle;
  weight?: "bold" | "bolder" | "lighter" | "normal" | FontWeight;
  decoration?: "line-through" | "underline" | "none" | FontDecoration;
}
```

Can we check the impact of doc in both cases? 

Object Literal approach preserves the backwards compatibility and provides type for uses who would like to use `FontStyle.Normal`.

On the question of if we choose `enum` vs `const enum` const enum will not add extra lookup table and it is better for smaller bundle sizes but we loose iterating over the values.

On using `const enum vs Object Literal`, the bundle size will increase with `Object as Const` approach as we  are creating an Object. `Const Enum` gets resolved at compile time itself.

Overall the merged PR does not have a breaking change , the issue would be that we will have both string union types and enum, this might confuse the user. Other than that it is a good solution.

Since JS does not have native enum support, there is a [stage 0 proposal](https://github.com/rbuckton/proposal-enum)l which may deviate from TS enums.
So there is group of users who suggest to avoid Enums and use alternatives like `Object Literal as const` approaches. The Pro Enum group advocate [string based enums are good](https://2ality.com/2020/01/typescript-enums.html#recommendation%3A-prefer-string-based-enums).

I believe it boils down to on the API if we want to use enums or alternatives to enums to provide Types instead of string union types.

If we want to decide purely on bundle size, string based [const enums ](https://2ality.com/2020/01/typescript-enums.html#const-enums) are good though we miss capability to iterate over possible values. 

If we want the iteration over possible values `Object based approach as const` is better alternative than [producing lookup table](https://netbasal.com/a-smaller-bundle-with-const-enums-in-typescript-b7f786227e5e) with using `enum`.

Comparisons of various [approaches](https://2ality.com/2020/02/enum-alternatives-typescript.html#summary-of-enums-and-enum-alternatives) 

cc: @dasa

_Originally posted by @subgan82 in https://github.com/Esri/arcgis-rest-js/pull/656#issuecomment-592543437_

---

We do use const enum, but we can't use `as const` yet since we're still on TS 3.3, but that does seem to be the preferred way going forward unless you need to performance of const enum and don't need the runtime inspection.

_Originally posted by @dasa in https://github.com/Esri/arcgis-rest-js/pull/656#issuecomment-592577726_

**Comment by subgan82** [comment]
cc: @maneetgoyal @dasa @tomwayson

**Comment by subgan82** [comment]
I have a branch with this approach, will push it after making some changes.

**Comment by patrickarlt** [comment]
I'm finally catching up with this. But i have some pretty strong opinions on this readying through your research @subgan82.

> If we want to decide purely on bundle size, string based const enums are good though we miss capability to iterate over possible values.

> If we want the iteration over possible values Object based approach as const is better alternative than producing lookup table with using enum.

While I understand that we might want to iterate over possible enums i don't think it is the job of ArcGIS REST JS to provide that. We discussed something similar to this in https://github.com/Esri/arcgis-rest-js/issues/12#issue-248271081 when we were spec'ing our how much of the API to document with TypeScript types. The general consensus was that we should have basic types available for request options and response properties for TypeScript users. I would argue that having iterable enums for use in things like building form dropdowns is outside the scope of ArcGIS REST JS.

**tldr;**

* Outside scope of ArcGIS REST JS. This is a library for the REST API not a giant shared type repository, we write type to make the API better for TS users and for documentation, not in and of themselves.
* Bloats bundle sizes. Small bundles are a key feature of ArcGIS REST JS. The object based approach would bloat bundle sizes and the objects cant be treeshaken out because methods depend on them.

cc: @maneetgoyal @dasa @tomwayson

**Comment by subgan82** [comment]
Thanks @patrickarlt for your feedback. We did update last [release](https://github.com/Esri/arcgis-rest-js/pull/656) to use `const enums ` along with `string literal union types` that is backwards compatible. 

In that we had discussions on benefits of using `Object Literals as Const` approach if we need to bring the iterable enums. Yes it has its pros/cons.

The `arcgis-rest-js` [types](https://devtopia.esri.com/WebGIS/arcgis-charts/blob/master/packages/arcgis-charts-spec/src/web-chart.ts#L9) are very useful in projects like ArcGIS Charts where it is used in the charts specification to represent the REST types. Even in this usage we found it useful to use Enums when implementing the charting [library](https://devtopia.esri.com/WebGIS/arcgis-charts/blob/master/packages/arcgis-charts-js/src/chart-library-elements/symbols/index.ts#L8). Enums provide better type support over string literals. What we have in the current release of arcgis-rest-js is good with const enums and string literal types. There are some advantages when using Object literals.

Do we envision a typescript library with all the rest types used with different TS projects  within esri ? you had mentioned about it in [here]( think we should author typescript types for requests and responses. @esri/rest-common-types)

Otherwise based on the current project goals of arcgis-rest-js we can defer this request to Object based Enums.

**Comment by ssylvia** [comment]
@patrickarlt I just updated to the most recent version of the app and noticed some breaking changes to the imports we are using from the `@esri/arcgis-rest-types`, specifically the `FieldType` import. It looks like these were converted over to the `const enums` in this commit as part of 2.9: https://github.com/Esri/arcgis-rest-js/commit/fbce4e8e1eacdc0a1a53be670dadf0f0bdc39d80. 

`const enums` require types to compile correctly and are not compatible when transpiling typescript with Babel

**Comment by tomwayson** [comment]
I think we should revert those changes and cut a patch release.

**Comment by subgan82** [comment]
Hi @ssylvia 

Can you use [babel-plugin-const-enum](https://www.npmjs.com/package/babel-plugin-const-enum)? 

https://babeljs.io/docs/en/babel-plugin-transform-typescript, The docs mentions this plugin as a workaround.

Babel has feature request for this [support](https://github.com/babel/babel/issues/8741) support.

**Comment by ssylvia** [comment]
@subgan82 I did try install that and running the build but received the same error. Babel doesn't compile node_module files

**Comment by subgan82** [comment]
> I think we should revert those changes and cut a patch release.

This could potentially break any package using Enums after the 2.9 update. In that commit we added const enum for these types and kept the string literals also. Apart from this one use-case we can roll back.

@dasa , @patrickarlt  your thoughts/feedback on this issue.

**Comment by patrickarlt** [comment]
I agree with @tomwayson I think we should revert https://github.com/Esri/arcgis-rest-js/commit/fbce4e8e1eacdc0a1a53be670dadf0f0bdc39d80 and cut a new patch. The changes introduced in 2.9 resulted in bugs and should be reverted if anyone is relying on them they are effectively relying on a bug.

**Comment by subgan82** [comment]
is adding --isolatedModules option to TSC useful to catch such errors in future.
 https://www.typescriptlang.org/docs/handbook/compiler-options.html

```
--isolatedModules | boolean | false | Perform additional checks to ensure that separate compilation (such as with transpileModule or @babel/plugin-transform-typescript) would be safe.
-- | -- | -- | --
```

**Comment by subgan82** [comment]
Added a PR to revert the const enum change

**Comment by subgan82** [comment]
using `isoloateModules: true` requires to make some more changes in tsconfig.json like `declaration` cannot be used with this option , so didn't make any changes for this in that PR.

**Comment by subgan82** [comment]
Using `as const typed objects` would work for babel scenarios and provide enum like support,  but based on this package [goals](https://github.com/Esri/arcgis-rest-js/issues/668#issuecomment-605080284) we cannot use it.

**Comment by subgan82** [comment]
These are the 3 files which are using union of string literals now, added a size comparison table 


File | Using string literal union types | Using Object as Const + string literal union types | Percentage increase
-- | -- | -- | --
statisisticDefinition.d.ts | 1082 | 1718 | 58%
symbol.d.ts | 2767 | 4753 | 71%
webmap.d.ts | 61786 | 63159 | 2%

**Comment by maneetgoyal** [comment]
> @subgan82 I did try install that and running the build but received the same error. Babel doesn't compile node_module files

@ssylvia Babel doesn't compile `node_modules` by default but it can be configured to compile it by updating the `ignore` option in Babel config.

> The changes introduced in 2.9 resulted in bugs and should be reverted if anyone is relying on them they are effectively relying on a bug.

@patrickarlt Apart from the babel compilation error, were there other bugs too that were reported? If just Babel compilation is an issue, configuring `babel.config.js` should solve it.

**Comment by patrickarlt** [comment]
> @ssylvia Babel doesn't compile node_modules by default but it can be configured to compile it by updating the ignore option in Babel config.

Having done this or tried to do this, this almost always results in more problems. Most modules assume they WON'T get recompiled and this tends to cause problems.

> @patrickarlt Apart from the babel compilation error, were there other bugs too that were reported? If just Babel compilation is an issue, configuring babel.config.js should solve it.

No but Babel is something we try to make sure works REALLY well out of the box. I don't want to force users to do odd things to make this library work under any circumstance.

@subgan82 Needing to iterate over types seems like a really.  specific use case for 1 product (ArcGIS Charts) right now. There are a lot of reasons I would rather not use `as const type objects` or `Object Literal as const`:

1. Only 1 product needs it, ArcGIS Charts
2. No other products or users seem to need it.
3. The functionality is easily replicated in ArcGIS Charts (Just write these types in ArcGIS Charts)
4.  The solution bloat bundle size significantly when it is applied. (https://github.com/Esri/arcgis-rest-js/issues/668#issuecomment-607970346)
5. The solution introduces odd errors and workarounds. (https://github.com/Esri/arcgis-rest-js/issues/668#issuecomment-607984377)
6. The solution is unevenly applied if SOME types are iterable but not others, the library the library becomes harder to use and its behavior becomes harder to predict.
7. This massively increases the scope of ArcGIS REST JS to include becoming an authoritative type library that can be consumed by other libraries.

For these reasons I would really like to just merge https://github.com/Esri/arcgis-rest-js/pull/681 and close this issue.

**Comment by subgan82** [comment]
> For these reasons I would really like to just merge #681 and close this issue.

👍
